[WIP] Rename object in events' parameters

Quarkstar opened the pull request:
Fix #1351 
### Screenshots
![fix](https://user-images.githubusercontent.com/32167255/79201765-21afef00-7e6b-11ea-86c4-f677b3d3d153.gif)

### Questions

#### `somethingModified` 
This is in  all rename functions in `Core/GDCore/IDE/Events/EventsRefactorer.cpp`, such as `RenameObjectInActions`.[(here)](https://github.com/4ian/GDevelop/blob/804a07c56ef756f078f24c689c6d5903d50ea6c1/Core/GDCore/IDE/Events/EventsRefactorer.cpp#L177)  
It seems like it's always false. :joy: I don't find which line to change it to true.

####  Const
`vector<pair<const gd::Expression*, const gd::ParameterMetadata>> GetAllExpressionsWithMetadata() const;`

I add the above function, just like what other `rename functions` do.  But it doesn't seem to be called. Is this useful?

#### Parameter Type
For repeat event, I use a metadata of `expression`. For foreach event, I use a metadata of `object`.
Is there any other conditions?

4ian commented:
That seems mostly good! I'll double check the code tomorrow :) 

> For repeat event, I use a metadata of expression. For foreach event, I use a metadata of object.
> Is there any other conditions?

There is also the JsCodeEvent but the objects are stored in a gd::String and not a gd::Expression.  We should change this but this can be done in another PR if you prefer because you also then need to update the JavaScript code that are calling getParameterObjects: https://github.com/4ian/GDevelop/search?q=GetParameterObjects&unscoped_q=GetParameterObjects.

> I add the above function, just like what other rename functions do. But it doesn't seem to be called. Is this useful?

The `const` version of a method is called when you call the method on an object that is itself const.
For example:

```
const MyClass & myConstObject = ...;
MyClass myObject = ...;

myObject.method(); // Call the non const
myConstObject.method(); // Call the const version
````

In practice it may not be needed now.. but is still a good practice to provide one in case something in the future want to analyze the events without modifying it (in which case we want to return the vector but with const gd::Expression*).

> It seems like it's always false. 😂

Might be a mistake on my side! I'll double check :D 

Thanks!

4ian commented on Core/GDCore/Events/Builtin/ForEachEvent.cpp:
```diff
@@ -37,6 +37,18 @@ vector<gd::Expression*> ForEachEvent::GetAllExpressions() {
+    ForEachEvent::GetAllExpressionsWithMetadata() {
+  vector<pair<gd::Expression*, gd::ParameterMetadata> >
+      allExpressionsWithMetadata;
+  auto type = gd::ParameterMetadata().SetType("object");
```
nitpicking: I would name this `metadata` or `parameterMetadata` rather than type

4ian commented on Core/GDCore/Events/Builtin/ForEachEvent.cpp:
```diff
@@ -59,6 +71,17 @@ vector<const gd::Expression*> ForEachEvent::GetAllExpressions() const {
   return allExpressions;
 }
 
+vector<pair<const gd::Expression*, const gd::ParameterMetadata> >
```
nitpicking: put the implementation of the const version immediately after the non const implementation

4ian replied:
Actually no, it's not how it was done in this file. Forget this! :)

4ian commented on Core/GDCore/Events/Event.h:
```diff
@@ -139,6 +140,21 @@ class GD_CORE_API BaseEvent {
   };
 
+  /**
+   * \brief Return a list of all expressions with ParameterMetadata of the event.
```
Return a list of all expressions of the event, each with their associated metadata.

4ian commented on Core/GDCore/IDE/Events/EventsRefactorer.h:
```diff
@@ -146,6 +147,19 @@ class GD_CORE_API EventsRefactorer {
                                        gd::String oldName,
                                        gd::String newName);
+   /**
+   * Replace all occurences of an object name by another name in a parameter of event.
```
Actually this function is doing a job not tied to an event:

`Replace all occurrences of an object name by another name in an expression with the specified metadata.`

4ian commented:
So I have mostly nitpciking. This seems great! 
Three things:
* `somethingModified` is for now useless, because not exposed back to the caller of RenameObjectInEvents. No big deal, *let's keep it like this* and fix that in another PR maybe if it's useful for the editor.
* Let's do the JSCodeEvent.h in another PR.
* Something to consider now: can you verify if GetAllExpressions() is even used? You added GetAllExpressionsWithMetadata - which is great. But it seems that `GetAllExpressions()` is already there and is kind of useless. I would say: remove it (in all events inheriting from `gd::BaseEvent`). Either
  * It will still compile, i.e: it was dead code.
  * Or it was not dead code and used somewhere => inspect if we can use `GetAllExpressionsWithMetadata()` instead, so that there is only the need for `GetAllExpressions()`

After that we should be good to go!

Quarkstar commented:
Thank you! I have fixed them all.

`GetAllExpressions` is useless and I removed it. Now there is only `PlatformExtension::GetAllExpressions` in Core.

Quarkstar commented:
Sorry for that, I forgot them. :cry:

4ian commented:
That's alright, I will merge once tests are passing :)

Quarkstar commented:
Thank you! I will be more careful next time.

4ian commented:
Done! Thanks for working on this 👍
